Math/NumberTheory/DirichletCharacters.hs

```diff
@@ -0,0 +1,589 @@
+                     then Just $ RealChar $ Generated $ map go $ snd $ mkTemplate n
+                     else Nothing
+  where n = natVal (Proxy :: Proxy n)
+        go :: Template -> DirichletFactor
+        go (OddTemplate p k g _) = OddPrime p k g $ toRootOfUnity (k % 2)
+          -- jacobi symbol of a primitive root mod p over p is always -1
+        go _ = error "internal error in jacobiCharacter: please report this as a bug"
+          -- every factor of n should be odd
+
+-- | A Dirichlet character is real if it is real-valued.
+newtype RealCharacter n = RealChar { -- | Extract the character itself from a `RealCharacter`.
+                                     getRealChar :: DirichletCharacter n
+                                   }
+                                   deriving Eq
+
+-- | Test if a given `DirichletCharacter` is real, and if so give a `RealCharacter`.
+isRealCharacter :: DirichletCharacter n -> Maybe (RealCharacter n)
+isRealCharacter t@(Generated xs) = if all real xs then Just (RealChar t) else Nothing
+  where real :: DirichletFactor -> Bool
+        real (OddPrime _ _ _ a) = a <> a == mempty
+        real (TwoPower _ _ b) = b <> b == mempty
+        real Two = True
+
+-- TODO: it should be possible to calculate this without evaluate/generalEval
+-- and thus avoid using discrete log calculations: consider the order of m
+-- inside each of the factor groups?
+-- | Evaluate a real Dirichlet character, which can only take values \(-1,0,1\).
+toRealFunction :: (Integral a, KnownNat n) => RealCharacter n -> a -> Int
+toRealFunction (RealChar chi) m = case generalEval chi (fromIntegral m) of
+                                    Zero -> 0
+                                    NonZero t | t == mempty -> 1
+                                    NonZero t | t == RootOfUnity (1 % 2) -> -1
+                                    _ -> error "internal error in toRealFunction: please report this as a bug"
+                                      -- A real character should not be able to evaluate to
+                                      -- anything other than {-1,0,1}, so should not reach this branch
+
+-- | Test if the internal DirichletCharacter structure is valid.
+validChar :: forall n. KnownNat n => DirichletCharacter n -> Bool
+validChar (Generated xs) = correctDecomposition && all correctPrimitiveRoot xs && all validValued xs
+  where correctDecomposition = sort (factorise n) == map getPP xs
+        getPP (TwoPower k _ _) = (two, fromIntegral k)
+        getPP (OddPrime p k _ _) = (p, k)
+        getPP Two = (two,1)
+        correctPrimitiveRoot (OddPrime p k g _) = g == generator p k
+        correctPrimitiveRoot _ = True
+        validValued (TwoPower k a b) = a <> a == mempty && (bit (k-2) :: Integer) `stimes` b == mempty
+        validValued (OddPrime (unPrime -> p) k _ a) = (p^(k-1)*(p-1)) `stimes` a == mempty
+        validValued Two = True
+        n = natVal (Proxy :: Proxy n)
+        two = toEnum 1 -- lazy way to get Prime 2
```

Comment:
Better use `minBound` :) 
It is actually an oversight from my side that `toEnum 1` is 2 and not 3.

Reply:
`minBound` doesn't work here because we would like a `Prime Natural`, which is not `Bounded`.

Reply:
Ahh, I see. What about `nextPrime 2`?

Reply:
I switched to this, since I had no strong preference either way.